server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -308,26 +342,126 @@ private void takeSnapshot(Map<String, LookupExtractorFactoryContainer> lookupMap
-      ImmutableMap.Builder<String, LookupExtractorFactoryContainer> builder = ImmutableMap.builder();
-
-      final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
-      for (LookupBean lookupBean : lookupBeanList) {
-        LookupExtractorFactoryContainer container = lookupBean.getContainer();
+    if (!lookupConfig.getDisableLookupSyncOnStartup()) {
+      String tier = lookupListeningAnnouncerConfig.getLookupTier();
+      List<LookupBean> lookupBeanList = getLookupListFromCoordinator(tier);
+      if (lookupBeanList == null) {
+        LOG.info("Coordinator is unavailable. Loading saved snapshot instead");
+        lookupBeanList = getLookupListFromSnapshot();
+      }
+      if (lookupBeanList != null) {
+        ImmutableMap.Builder<String, LookupExtractorFactoryContainer> builder = ImmutableMap.builder();
+        ListeningScheduledExecutorService executorService = MoreExecutors.listeningDecorator(
+            Executors.newScheduledThreadPool(
+                lookupConfig.getNumLookupLoadingThreads(),
+                Execs.makeThreadFactory("LookupReferencesManager-Startup-%s")
+            )
+        );
+        try {
+          List<ListenableFuture<Map.Entry>> futures = new ArrayList<>();
+          LOG.info("Starting lookup loading process");
+          for (LookupBean lookupBean : lookupBeanList) {
+            futures.add(
+                executorService.submit(
+                    () -> {
+
+                      LookupExtractorFactoryContainer container = lookupBean.getContainer();
+                      LOG.info(
+                          "Started lookup [%s]:[%s]",
+                          lookupBean.getName(),
+                          container
+                      );
+                      if (container.getLookupExtractorFactory().start()) {
+                        return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
+                      } else {
+                        LOG.error(
+                            "Failed to start lookup [%s]:[%s]",
+                            lookupBean.getName(),
+                            container
+                        );
+                        return null;
+                      }
+                    }
+                )
+            );
+          }
+          final ListenableFuture<List<Map.Entry>> futureList = Futures.allAsList(futures);
+          try {
+            futureList.get()
+                      .stream()
+                      .filter(Objects::nonNull)
+                      .forEach(builder::put);
+          }
+          catch (Exception ex) {
+            futureList.cancel(true);
+            throw ex;
+          }
+          stateRef.set(new LookupUpdateState(builder.build(), ImmutableList.of(), ImmutableList.of()));
+        }
+        catch (Exception e) {
+          LOG.error(e, "Failed to finish lookup load process.");
+        }
+        executorService.shutdownNow();
+      } else {
```

Comment:
can you also wait for executorService to finish using executorService.awaitTermination(1 minute) ... if it doesn't finish within that duration then log a warning.